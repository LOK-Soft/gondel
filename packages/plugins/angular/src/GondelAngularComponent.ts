import { GondelBaseComponent } from "@gondel/core";
import { PlatformRef, StaticProvider, Type, NgModuleRef } from "@angular/core";
import { platformBrowserDynamic } from "@angular/platform-browser-dynamic";
import { NgModuleFactory } from "@angular/core/src/render3";
import { BootstrapOptions } from "@angular/core/src/application_ref";

export class GondelAngularComponent extends GondelBaseComponent {
  public static globalPlatformRef: PlatformRef;

  protected bootstrapModuleFactory<T extends {}>(
    extraProviders: StaticProvider[] | undefined,
    moduleFactory: NgModuleFactory<{}>,
    options?: BootstrapOptions | undefined
  ): Promise<NgModuleRef<T>> {
    return new Promise((resolve, reject) => {
      platformBrowserDynamic(extraProviders)
        .bootstrapModuleFactory(moduleFactory, options)
        .then((factory: NgModuleRef<{}>) => {
          resolve(factory as NgModuleRef<any>);
        })
        .catch(reject);
    });
  }

  protected bootstrapModule<T extends {}>(
    extraProviders: StaticProvider[] | undefined,
    rawModule: Type<{}>,
    options?: BootstrapOptions | undefined
  ): Promise<NgModuleRef<T>> {
    return new Promise((resolve, reject) => {
      platformBrowserDynamic(extraProviders)
        .bootstrapModule(rawModule, options)
        .then((factory: NgModuleRef<{}>) => {
          resolve(factory as NgModuleRef<any>);
        })
        .catch(reject);
    });
  }
}
