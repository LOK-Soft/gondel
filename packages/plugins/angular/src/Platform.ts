import { PlatformRef, StaticProvider, NgModuleRef } from "@angular/core/";
import { platformBrowserDynamic } from "@angular/platform-browser-dynamic";
import { NgModuleFactory } from "@angular/core/src/render3";

enum PlatformDefs {
  Dynamic
}

interface NgWindow {
  ngRef: NgModuleRef<any>;
}

export const hot = (boot: Promise<NgModuleFactory<any>>) => {
  boot.then((ref: NgModuleFactory<any>) => {
    // Ensure Angular destroys itself on hot reloads.
    const ngRef: NgModuleRef<any> | undefined = (window as any).ngRef;
    if (ngRef) {
      (ngRef as NgModuleRef<any>).destroy();
    }
    (window as any)["ngRef"] = ref;
  });
};

export const Platform: {
  [key in PlatformDefs]: (providers: StaticProvider[] | undefined) => PlatformRef
} = {
  [PlatformDefs.Dynamic]: platformBrowserDynamic
};
