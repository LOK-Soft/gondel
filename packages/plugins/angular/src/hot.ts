import { NgModuleRef } from "@angular/core";
import { NgModuleFactory } from "@angular/core/src/render3";

export const hot = (ref: NgModuleFactory<any>) => {
  // ensure Angular destroys itself on hot reloads.
  const ngRef: NgModuleRef<any> | undefined = (window as any).ngRef;
  if (ngRef) {
    (ngRef as NgModuleRef<any>).destroy();
  }
  (window as any)["ngRef"] = ref;
};
