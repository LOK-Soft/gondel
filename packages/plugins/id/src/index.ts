/**
 * This is a demo plugin which adds a data-t-id for demonstration purposes
 */
import { addGondelPluginEventListener, GondelComponent } from "@gondel/core";

export function init(namespace: string) {
  let id = 0;
  addGondelPluginEventListener("start", function addDataId(
    components: Array<GondelComponent>,
    data,
    resolve
  ) {
    if (data.namespace !== namespace) {
      return resolve(components);
    }
    components.forEach(component => {
      component._ctx.setAttribute(`data-${namespace}-id`, (id++).toString());
    });
    resolve(components);
  });
}
