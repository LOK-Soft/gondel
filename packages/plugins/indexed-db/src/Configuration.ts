export let VERSION = 1;
