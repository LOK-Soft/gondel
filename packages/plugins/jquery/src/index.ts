/**
 * This is a demo plugin which adds a _$ctx for demonstration purposes
 */
import { GondelComponent, addGondelPluginEventListener } from "@gondel/core";
import $ from "jquery";

export function init() {
  addGondelPluginEventListener("start", function addJquery(
    components: Array<GondelComponent>,
    data: any,
    resolve: (components: Array<GondelComponent>) => void
  ) {
    components.forEach(component => {
      (component as any).$ctx = $(component._ctx);
    });
    resolve(components);
  });
}
