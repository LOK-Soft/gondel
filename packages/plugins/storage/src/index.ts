import { Adapter, IStorage, ISerializer } from "./Adapter";
import { storage, setDefaultStorageAdapter } from "./StorageDecorator";
import BooleanSerializer from "./serializer/Boolean";
import NumberSerializer from "./serializer/Number";
import JSONSerializer from "./serializer/JSON";

export {
  Adapter,
  storage,
  BooleanSerializer,
  NumberSerializer,
  JSONSerializer,
  setDefaultStorageAdapter,
  IStorage,
  ISerializer
};
