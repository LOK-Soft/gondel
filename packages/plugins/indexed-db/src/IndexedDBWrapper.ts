import { IndexedDB, Transaction } from "./NativeIndexedDB";
import { VERSION } from "./Configuration";

export interface IIndexedDBEntry {
  id: string;
}

class IndexedDBWrapperError extends Error {
  constructor(method: string, message?: string) {
    super(`Transaction:${method}: ${message || "unknown error occured"}`);
  }
}

const throwOnInvalidConnection = (
  database: IDBDatabase | undefined,
  method: string,
  reject: (err: IndexedDBWrapperError) => void
) => {
  if (!database) {
    reject(
      new IndexedDBWrapperError(
        method,
        `connection is not established, probably connect() call missing`
      )
    );
  }
};

export class IndexedDBWrapper {
  private database: IDBDatabase;

  constructor(private name: string, private version: number = VERSION) {}

  public async connect(): Promise<any> {
    return this.establishInternalConnection();
  }

  public get<T extends Object>(idOrEntry: string | IIndexedDBEntry): Promise<T & IIndexedDBEntry> {
    return new Promise((resolve, reject) => {
      console.log("get", this.database);
      throwOnInvalidConnection(this.database, "get", reject);

      const transaction: IDBRequest = this.database
        .transaction([this.name])
        .objectStore(this.name)
        .get(typeof idOrEntry === "string" ? idOrEntry : idOrEntry.id);

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "get",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };

      transaction.onsuccess = () => {
        resolve(transaction.result as T & IIndexedDBEntry);
      };
    });
  }

  public set<T extends IIndexedDBEntry>(data: T): Promise<void> {
    return new Promise((resolve, reject) => {
      console.log("set", this.database);
      throwOnInvalidConnection(this.database, "set", reject);

      const transaction = this.database
        .transaction([this.name], "readwrite")
        .objectStore(this.name)
        .add(data);

      transaction.onsuccess = () => {
        resolve();
      };

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "set",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };
    });
  }

  public delete(idOrEntry: string | IIndexedDBEntry): Promise<void> {
    return new Promise((resolve, reject) => {
      throwOnInvalidConnection(this.database, "delete", reject);
      const transaction = this.database
        .transaction([this.name], "readwrite")
        .objectStore(this.name)
        .delete(typeof idOrEntry === "string" ? idOrEntry : idOrEntry.id);

      transaction.onsuccess = () => {
        resolve();
      };

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "delete",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };
    });
  }

  public __dangerouslyAccessInternalDatabase(): IDBDatabase {
    return this.database;
  }

  private establishInternalConnection(): Promise<IDBDatabase> {
    return new Promise((resolve, reject) => {
      const self = this;
      const transaction = IndexedDB.open(this.name, this.version);

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "connect",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };

      transaction.onsuccess = (event: Event) => {
        this.database = transaction.result;

        resolve(this.database);
      };

      transaction.onupgradeneeded = function(event: IDBVersionChangeEvent) {
        if (event.newVersion && event.oldVersion < event.newVersion) {
          ((event.target as any).result as IDBDatabase).createObjectStore(self.name, {
            keyPath: "id"
          });
        }
      };
    });
  }
}
