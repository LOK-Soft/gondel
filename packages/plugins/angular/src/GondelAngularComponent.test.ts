import { GondelComponent, Component, startComponents, getComponentByDomNode } from "@gondel/core";
import { GondelAngularComponent } from "./GondelAngularComponent";
import { NgModuleFactory } from "@angular/core/src/render3";
import { NgModuleRef } from "@angular/core";

function createMockElement<T extends GondelComponent>(namespace: string = "g") {
  const demoElement = document.createElement("div");
  demoElement.innerHTML = `<p>Loading...</p>`;
  demoElement.setAttribute("data-" + namespace + "-name", "DemoComponent");
  document.documentElement!.appendChild(demoElement);
  startComponents(demoElement, namespace);
  return getComponentByDomNode<T>(demoElement, namespace)!;
}

describe("@gondel/plugin-angular", () => {
  it("should boot a single component without crashing", () => {
    @Component("DemoComponent")
    class DemoComponent extends GondelAngularComponent {
      public ref: NgModuleRef<any>;

      async load() {
        const {
          AppModule
        }: {
          AppModule: NgModuleFactory<{}>;
        } = await import(/* webpackChunkName: "demoModule" */ "../test/fixtures/app/dist/ngc/src/app/app.module.ngfactory.js");
        return AppModule;
      }

      async start() {
        const AppModule = await this.load();
        this.ref = await this.bootstrapModuleFactory(undefined, AppModule);
      }
    }

    const component = createMockElement<DemoComponent>("g");
    expect(component).toMatchSnapshot();
    expect(component.ref).toBeDefined();
    expect(component.ref).toMatchSnapshot();
  });
});
