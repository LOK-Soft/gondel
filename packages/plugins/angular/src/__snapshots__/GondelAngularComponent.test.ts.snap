// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@gondel/plugin-angular should boot a single component without crashing 1`] = `
DemoComponent {
  "_componentName": "DemoComponent",
  "_ctx": <div
    data-g-name="DemoComponent"
  >
    <p>
      Loading...
    </p>
  </div>,
  "_namespace": "g",
  "stop": [Function],
}
`;
